crates/sui-core/src/authority.rs

```diff
@@ -2317,18 +2365,18 @@ impl AuthorityState {
                     "handle_consensus_transaction UserTransaction",
                 );
 
-                // Schedule the certificate for execution
-                self.add_pending_sequenced_certificate(certificate.clone())?;
-
                 if certificate.contains_shared_object() {
                     self.database
-                        .lock_shared_objects(&certificate, consensus_index)
-                        .await
+                        .record_shared_object_cert_from_consensus(&certificate, consensus_index)
+                        .await?;
                 } else {
                     self.database
                         .record_owned_object_cert_from_consensus(&certificate, consensus_index)
-                        .await
+                        .await?;
                 }
+
+                let mut transaction_manager = self.transaction_manager.lock().await;
+                transaction_manager.enqueue(vec![certificate]).await
```

Comment:
Question about this (apologies if this was answered somewhere else). I see that transaction_manager seem to be in-memory data structure. What happens if node restarts? Who guarantees that transaction will be executed eventually?

Reply:
I mean they can be think of as the replay log for crash recovery of the TransactionManager, even though it is technically a map / rocksdb table. Basically we re-apply pending certificates on restart.

Reply:
Sorry, I still don't fully follow. `TransactionManager::pending_certificates` is a BTreeMap, how can it be used to re-apply certificates on restart? Is the plan to change it to DBMap in the future?

Reply:
Got it. And I see that we now write to `pending_certificates` in `finish_consensus_message_process`. Thanks for clarification!